File: code/java/concurrency/02.java
public &ltT&gt CompletableFuture&ltT&gt inTransaction(DatabaseFunction&ltT&gt fun) {
    return CompletableFuture
        .runAsync(() -> this.sendQuery("BEGIN"))
        .thenApply(ignored -> fun.execute(this))
        .thenApply(s -> {
            this.sendQuery("COMMIT");
            return s;
        }).exceptionally(throwable -> {
            this.sendQuery("ROLLBACK");
            return null;
        });
}

File: code/kotlin/concurrency/02.kt
suspend fun &ltT&gt inTransaction(
    f: suspend (Connection) -> T): T {
    try {
        this.sendQuery("BEGIN")
        val result = f(this)
        this.sendQuery("COMMIT")
        return result
    } catch (e: Throwable) {
        this.sendQuery("ROLLBACK")
        throw e
    }
}

